tree/tree/src/TTree.cxx

```diff
@@ -1209,6 +1209,19 @@ void TTree::AddClone(TTree* clone)
    }
 }
 
+void ThrowIfReshuffled(TTree *t1, TTree *t2)
+{
+   const auto isT1Reshuffled = t1->TestBit(TTree::kEntriesReshuffled);
+   const auto isT2Reshuffled = t2->TestBit(TTree::kEntriesReshuffled);
+   if (isT1Reshuffled || isT2Reshuffled) {
```

Comment:
This might be restrictive.   As even after reshuffling, if the t2 has a TTreeIndex that is compatible with t1 (i.e. the branches used for the index are present in t1), in this case the result will be correct even if the performance will be degraded (friend read in non-linear order)

Reply:
The issue is not about branches being present/absent, but with mismatched event ordering.

Even if there is a TTreeIndex, how can we know if it refers to the correct, reshuffled event ordering? Given that the only facility that sets the `kEntriesReshuffled` bit in ROOT is `Snapshot`, and RDF cannot produce TTreeIndexes, it's likely that a `TTreeIndex` does not correspond to a TTree with kEntriesReshuffled set.

With the reasoning above, plus the relaxation from `throw` to `Warn`/`Error`, do you think we can leave the check as it is?

Reply:
> but with mismatched event ordering.

This is **precisely** the challenge that the TTreeIndex addresses.   Upon loading an entry, the main TTree reads one or two values (eg. Run#, Event#) and look them up in the friend tree's TTreeIndex which tells it which of its entry correspond to the main tree entry being read.

> how can we know if it refers to the correct, reshuffled event ordering?

If the main tree has the branch (name) used for building the friend tree index then we can assume that the information is provided by the index (of course we can not protect against the user changing the semantic of the branch (name)).

> and RDF cannot produce TTreeIndexes, 

Then can trivially be created after the fact (or even just before running RDF).

[As a related aside Andrew M. just reminded me that CMS have the same problem (entry being reshuffled on processing ; seen in case where they produce a new (to be friended) tree containing new values to be associated with an existing tree (aka corrections) and the processing is done in multi-threaded way) and uses the index to solve it]

Reply:
> it's likely that a TTreeIndex does not correspond to a TTree with kEntriesReshuffled set.

Since the index is about the TTree (produced by the Snapshot) and using information only from that TTree, it can not possibly not correspond to it :)

Reply:
I can't say I'm clear on the details, but I see much more clearly than me on the topic. How would you update the check then?